datajunction-server/datajunction_server/api/nodes.py

```diff
@@ -459,6 +460,89 @@ def restore_node(name: str, *, session: Session = Depends(get_session)):
     )
 
 
+@router.delete(
+    "/nodes/{node_name}/hard/",
+    response_model=List[MaterializationConfigInfoUnified],
+)
+def hard_delete_a_node(
+    node_name: str,
+    session: Session = Depends(get_session),
+) -> JSONResponse:
+    """
+    Hard delete a node, destroying all links and invalidating all downstream nodes.
+    This should be used with caution, deactivating a node is preferred.
+    """
+    node = get_node_by_name(session, node_name, with_current=True)
+    downstream_nodes = get_downstream_nodes(session=session, node_name=node_name)
+
+    linked_nodes = []
+    if node.type == NodeType.DIMENSION:
+        linked_nodes = get_nodes_with_dimension(session=session, dimension_node=node)
+
+    session.delete(node)
+    session.commit()
```

Comment:
Should we do everything here in a single commit so that it can be rolled back together if one of these pieces doesn't succeed?

Reply:
So does that mean there isn't a way to do this all in one commit?

Reply:
I don't think there is a way, unless we can keep the node (being deleted) as existing but invalid and then re-validate the dag... and here comes the need to manual setting of a node to invalid. Something I never thought was needed.

Reply:
That's an interesting idea, we could deactivate the node, then revalidate all of the downstream and linked nodes to get the impact, and then delete the node. We can then catch any issues and back out by just reactivating the node. It all sounds like it would work theoretically but it feels like a lot of steps with gotchas that aren't obvious right now. 🤔

Reply:
Yes @samredai , I don't think we need to improve it right now, but what you said makes perfect sense as a final solution.

Reply:
+1, no need to change it now but we can get there eventually!